shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE PatternSynonyms #-}
+
+module Generator.Core
+  ( findPayKeyPair
+  , genTxOut
+  , genUtxo0
+  , traceKeyPairs
+  , someKeyPairs
+  , pickStakeKey
+  , genInteger
+  , genNatural
+  , toAddr
+  , toCred
+  , unsafeMkUnitInterval)
+  where
+
+import           Cardano.Crypto.DSIGN (deriveVerKeyDSIGN, genKeyDSIGN)
+import           Crypto.Random (drgNewTest, withDRG)
+import           Data.Maybe (fromMaybe)
+import           Data.Word (Word64)
+import           Hedgehog (Gen)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import           Numeric.Natural
+
+import           BaseTypes (UnitInterval, mkUnitInterval)
+import           Coin (Coin (..))
+import           Keys (pattern KeyPair, pattern SKey, pattern VKey, hashKey, vKey)
+import           LedgerState (genesisCoins)
+import           MockTypes (Addr, Credential, KeyPair, KeyPairs, SKey, TxOut, UTxO, VKey)
+import           Tx (pattern TxOut)
+import           TxData (pattern AddrBase, pattern KeyHashObj)
+
+mkKeyTuple :: (Word64, Word64, Word64, Word64, Word64) -> (VKey, SKey)
+mkKeyTuple seed = fst . withDRG (drgNewTest seed) $ do
+  sk <- genKeyDSIGN
+  return ( VKey $ deriveVerKeyDSIGN sk
+         , SKey sk)
+
+mkKeyPairs :: Word64 -> (KeyPair, KeyPair)
+mkKeyPairs n
+  = (mkKeyPair (2*n), mkKeyPair (2*n+1))
+  where
+    mkKeyPair n_ = (uncurry KeyPair) (mkKeyTuple (n_,n_,n_,n_,n_))
+
+-- | Constant list of KeyPairs intended to be used in the generators.
+traceKeyPairs :: KeyPairs
+traceKeyPairs = mkKeyPairs <$> [1 .. 200]
+
+-- | Select between _lower_ and _upper_ keys from 'traceKeyPairs'
+someKeyPairs :: Integer -> Integer -> Gen KeyPairs
+someKeyPairs lower upper = do
+  keys <- Gen.shuffle traceKeyPairs
+  n <- genInteger lower upper
+  return $ take (fromIntegral n) keys
```

Comment:
we might want to use `$!` so that the return result is in `whnf`.

Reply:
Here the choice of `keys` and `n` seem to be independent from each other. So we might want to do something like:

```haskell
(\keys n -> take (fromIntegral n) keys) <$> Gen.shuffle ... <*> genInteger ...
```

so that we get better shrinks.

Reply:
👍